src/main/scala/latis/input/HapiBinaryAdapter.scala

```diff
@@ -0,0 +1,128 @@
+package latis.input
+
+import java.net.URI
+import java.nio.charset.StandardCharsets
+import java.nio.{ByteBuffer, ByteOrder}
+
+import cats.effect.IO
+import fs2.Stream
+import latis.data.{Data, Sample}
+import latis.model.{DataType, Function, Scalar}
+
+/**
+ * Adapter for HAPI binary datasets.
+ */
+class HapiBinaryAdapter(model: DataType) extends StreamingAdapter[Array[Byte]] {
+  
+  val order = ByteOrder.LITTLE_ENDIAN //all numeric values are little endian (LSB)
+  
+  lazy val blockSize: Int = model.getScalars.map(bytesToRead).sum
+
+  /**
+   * Provides a Stream of records as Arrays of bytes.
+   */
+  def recordStream(uri: URI): Stream[IO, Array[Byte]] =
+    StreamSource.getStream(uri)
+      .chunkN(blockSize)
+      .map(_.toArray)
+  
+  /**
+   * Parses a record into a Sample. 
+   * Returns None if the record is invalid.
+   */
+  def parseRecord(record: Array[Byte]): Option[Sample] = {
+    // We assume one value for each scalar in the model.
+    // Note that Samples don't capture nested tuple structure.
+    // Assume uncurried model (no nested function), for now.
+    /*
+     * TODO: Traverse the model to build nested functions (Function in range) 
+     * should we apply currying logic or require operation? 
+     * curry should be cheap for ordered data
+     */
+
+    // Get Vectors of domain and range Scalar types.
+    val (dtypes, rtypes) = model match {
+      case Function(d, r) => (d.getScalars, r.getScalars)
+      case _ => (Vector.empty, model.getScalars)
+    }
+
+    // Extract the data values from the record
+    // and split into Vectors of domain and range.
+    val data = extractData(record)
+    val (ds, rs) = data.splitAt(dtypes.length)
+
+    // Construct a Sample from the domain and range values.
+    if (rtypes.length != rs.length) None //invalid record
+    else 
+      Some(Sample(ds, rs))
+  }
+
+  /**
+   * Extracts the data values from the given record as a List.
+   */
+  private def extractData(record: Array[Byte]): List[Data] = {
+    val buffer = ByteBuffer.wrap(record).order(order)
+    
+    model.getScalars.map { s => 
+      s("type") match {
+        case Some("short")  => Data(buffer.getDouble.toShort)
+        case Some("int")    => Data(buffer.getDouble.toInt)
+        case Some("long")   => Data(buffer.getDouble.toLong)
```

Comment:
The spec does say:

> integers are always signed and four byte

so I assume we want `getInt` instead of `getDouble` for these. It appears that "int", "double", and "string" are the only valid types we should get from HAPI.

Reply:
I interpreted that sentence differently and I know realize it's ambiguous. The full sentence is: "integers are always signed and four byte and floating point values are always IEEE 754 double precision values."

That can be read as either "_integers are always signed and four byte_, and floating point values are always IEEE 754 double precision values" (as you interpreted) or "_integers are always signed_, and four byte and floating point values are always IEEE 754 double precision values" (as I interpreted). I'm not sure which is the correct interpretation.

Reply:
Elsewhere in the spec it says:

> Binary content for double is always 8 bytes in IEEE 754 format, integer is 4 bytes signed little-endian.

Reply:
Okay! That solves it. Thanks again Chris. I'll update the code.